### PR TITLE
hpcgap: fix warnings in src/weakptr.c

### LINKED ISSUE #1278: Warning in weakptr.c in master branch (clang, hpcgap)
Another warning is here:
```
libtool: compile:  gcc -MQ obj/weakptr.lo -MMD -MP -MF obj/.deps/weakptr.d -m64 -Qunused-arguments -pthread -Werror -I./hpcgap -I. -DHAVE_CONFIG_H -I/Volumes/hudson-fs/workspace/GAP-newbuild-compilers/GAPCOPTS/64build/GAPREADLINE/noreadline/HPCMODE/hpcgap/label/clang-xcode/GAP-master-snapshot/extern/install/gmp/include -I/Volumes/hudson-fs/workspace/GAP-newbuild-compilers/GAPCOPTS/64build/GAPREADLINE/noreadline/HPCMODE/hpcgap/label/clang-xcode/GAP-master-snapshot/extern/install/gc/include -I/Volumes/hudson-fs/workspace/GAP-newbuild-compilers/GAPCOPTS/64build/GAPREADLINE/noreadline/HPCMODE/hpcgap/label/clang-xcode/GAP-master-snapshot/extern/install/libatomic_ops/include -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -Wno-missing-field-initializers -c gen/weakptr.c  -fno-common -DPIC -o obj/.libs/weakptr.o
hpcgap/src/weakptr.c:178:16: error: unused variable 'list2' [-Werror,-Wunused-variable]
  volatile Obj list2 = list;
               ^
hpcgap/src/weakptr.c:276:16: error: unused variable 'val2' [-Werror,-Wunused-variable]
  volatile Obj val2 = val;
               ^
2 errors generated.
make: *** [obj/weakptr.lo] Error 1
```
Happens when GAP is compiled with clang on macOS with hpcgap enabled.

### PR DESCRIPTION
**Pull request by fingolfin**
Resolves #1278 

@rbehrends please have a look at this, does it seem right?

### CODE REVIEW

File: src/weakptr.c [hunk 1]
```diff
@@ -163,20 +163,22 @@ Int GrowWPObj (
...
-Obj FuncWeakPointerObj( Obj self, Obj list ) { 
+#ifdef BOEHM_GC
+/* We need to make sure that <list> stays live until
+ * after REGISTER_WP(); on architectures that pass
+ * arguments in registers (x86_64, SPARC, etc), the
+ * argument register may be reused. In conjunction with
+ * loop unrolling, the reference to <list> may then be
+ * destroyed before REGISTER_WP() is called.
+ */
+Obj FuncWeakPointerObj( Obj self, volatile Obj list )
```
Comment (ChrisJefferson) [inline src/weakptr.c]: I'm unsure that this will do what we want -- does volatile on a function argument force it to be placed on the stack?

Comment (fingolfin) [inline src/weakptr.c]: Good question...

That said, I am not sure the previous code did what we want -- in particular, I didn't see anything in the C, see e.g. https://stackoverflow.com/questions/27816535

Maybe we need to rethink this completely...

Anyway, a "safer" fix than this PR might be to change the code to actually use `list2` -- in this function that should be easy.

Comment (ChrisJefferson) [inline src/weakptr.c]: My belief (which I have no evidence for) is adding `(void)list2;` to the old code will force `list2` to be considered "used", which should keep it present, and also stop the warning.

Comment (fingolfin) [inline src/weakptr.c]: The `(void)list2` trick is an old one, and not quite portable. In any case, I just rewrote this PR with a completely different approach.

**Pull request merged**